README
NAME

    Zing::Store::Pg - Postgres Storage

ABSTRACT

    Postgres Storage Abstraction

SYNOPSIS

      use Test::DB::Postgres;
      use Zing::Encoder::Dump;
      use Zing::Store::Pg;
    
      my $testdb = Test::DB::Postgres->new;
      my $store = Zing::Store::Pg->new(
        client => $testdb->create->dbh,
        encoder => Zing::Encoder::Dump->new
      );
    
      # $store->drop;

DESCRIPTION

    This package provides a Postgres-specific storage adapter for use with
    data persistence abstractions. The "client" attribute accepts a DBI
    object configured to connect to a DBD::Pg backend. The ZING_DBNAME
    environment variable can be used to specify the database name (defaults
    to "zing"). The ZING_DBHOST environment variable can be used to specify
    the database host (defaults to "localhost"). The ZING_DBPORT
    environment variable can be used to specify the database port (defaults
    to "5432"). The ZING_DBUSER environment variable can be used to specify
    the database username (defaults to "postgres"). The ZING_DBPASS
    environment variable can be used to specify the database password. The
    ZING_DBZONE environment variable can be used to specify the database
    table name (defaults to "entities").

INHERITS

    This package inherits behaviors from:

    Zing::Store

LIBRARIES

    This package uses type constraints from:

    Zing::Types

ATTRIBUTES

    This package has the following attributes:

 client

      client(InstanceOf["DBI::db"])

    This attribute is read-only, accepts (InstanceOf["DBI::db"]) values,
    and is optional.

METHODS

    This package implements the following methods:

 decode

      decode(Str $data) : HashRef

    The decode method decodes the JSON data provided and returns the data
    as a hashref.

    decode example #1

        # given: synopsis
      
        $store->decode('{"status"=>"ok"}');

 drop

      drop(Str $key) : Int

    The drop method removes (drops) the item from the datastore.

    drop example #1

        # given: synopsis
      
        $store->drop('zing:main:global:model:temp');

 encode

      encode(HashRef $data) : Str

    The encode method encodes and returns the data provided as JSON.

    encode example #1

        # given: synopsis
      
        $store->encode({ status => 'ok' });

 keys

      keys(Str @keys) : ArrayRef[Str]

    The keys method returns a list of keys under the namespace of the
    datastore or provided key.

    keys example #1

        # given: synopsis
      
        my $keys = $store->keys('zing:main:global:model:temp');

    keys example #2

        # given: synopsis
      
        $store->send('zing:main:global:model:temp', { status => 'ok' });
      
        my $keys = $store->keys('zing:main:global:model:temp');

 lpull

      lpull(Str $key) : Maybe[HashRef]

    The lpull method pops data off of the top of a list in the datastore.

    lpull example #1

        # given: synopsis
      
        $store->lpull('zing:main:global:model:items');

    lpull example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        $store->lpull('zing:main:global:model:items');

 lpush

      lpush(Str $key, HashRef $val) : Int

    The lpush method pushed data onto the top of a list in the datastore.

    lpush example #1

        # given: synopsis
      
        $store->lpush('zing:main:global:model:items', { status => '1' });

    lpush example #2

        # given: synopsis
      
        $store->lpush('zing:main:global:model:items', { status => '0' });

 recv

      recv(Str $key) : Maybe[HashRef]

    The recv method fetches and returns data from the datastore by its key.

    recv example #1

        # given: synopsis
      
        $store->recv('zing:main:global:model:temp');

    recv example #2

        # given: synopsis
      
        $store->send('zing:main:global:model:temp', { status => 'ok' });
      
        $store->recv('zing:main:global:model:temp');

 rpull

      rpull(Str $key) : Maybe[HashRef]

    The rpull method pops data off of the bottom of a list in the
    datastore.

    rpull example #1

        # given: synopsis
      
        $store->rpull('zing:main:global:model:items');

    rpull example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 1 });
        $store->rpush('zing:main:global:model:items', { status => 2 });
      
        $store->rpull('zing:main:global:model:items');

 rpush

      rpush(Str $key, HashRef $val) : Int

    The rpush method pushed data onto the bottom of a list in the
    datastore.

    rpush example #1

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });

    rpush example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });

 send

      send(Str $key, HashRef $val) : Str

    The send method commits data to the datastore with its key and returns
    truthy.

    send example #1

        # given: synopsis
      
        $store->send('zing:main:global:model:temp', { status => 'ok' });

 size

      size(Str $key) : Int

    The size method returns the size of a list in the datastore.

    size example #1

        # given: synopsis
      
        my $size = $store->size('zing:main:global:model:items');

    size example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        my $size = $store->size('zing:main:global:model:items');

 slot

      slot(Str $key, Int $pos) : Maybe[HashRef]

    The slot method returns the data from a list in the datastore by its
    index.

    slot example #1

        # given: synopsis
      
        my $model = $store->slot('zing:main:global:model:items', 0);

    slot example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        my $model = $store->slot('zing:main:global:model:items', 0);

 test

      test(Str $key) : Int

    The test method returns truthy if the specific key (or datastore)
    exists.

    test example #1

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        $store->test('zing:main:global:model:items');

    test example #2

        # given: synopsis
      
        $store->drop('zing:main:global:model:items');
      
        $store->test('zing:main:global:model:items');

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the terms of the The Apache License, Version 2.0, as elucidated in the
    "license file"
    <https://github.com/iamalnewkirk/zing-store-pg/blob/master/LICENSE>.

PROJECT

    Wiki <https://github.com/iamalnewkirk/zing-store-pg/wiki>

    Project <https://github.com/iamalnewkirk/zing-store-pg>

    Initiatives <https://github.com/iamalnewkirk/zing-store-pg/projects>

    Milestones <https://github.com/iamalnewkirk/zing-store-pg/milestones>

    Contributing
    <https://github.com/iamalnewkirk/zing-store-pg/blob/master/CONTRIBUTE.md>

    Issues <https://github.com/iamalnewkirk/zing-store-pg/issues>



CHANGES
Changelog for Zing-Store-Pg


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Postgres Storage Abstraction",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Zing-Store-Pg",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Zing::Store::Pg",
  "PREREQ_PM" => {
    "DBD::Pg" => "3.14.2",
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "Zing" => "0.20",
    "routines" => "0.02",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "DBD::Pg" => "3.14.2",
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "Test::Auto" => "0.12",
    "Test::DB" => "0.06",
    "Zing" => "0.20",
    "routines" => "0.02",
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBD::Pg" => "3.14.2",
  "Data::Object::Class" => "2.02",
  "Data::Object::ClassHas" => "2.01",
  "Test::Auto" => "0.12",
  "Test::DB" => "0.06",
  "Zing" => "0.20",
  "routines" => "0.02",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Zing/Store/Pg.pm
package Zing::Store::Pg;

use 5.014;

use strict;
use warnings;

use registry 'Zing::Types';
use routines;

use Data::Object::Class;
use Data::Object::ClassHas;

extends 'Zing::Store';

# VERSION

# ATTRIBUTES

has client => (
  is => 'ro',
  isa => 'InstanceOf["DBI::db"]',
  new => 1,
);

fun new_client($self) {
  my $dbname = $ENV{ZING_DBNAME} || 'zing';
  my $dbhost = $ENV{ZING_DBHOST} || 'localhost';
  my $dbport = $ENV{ZING_DBPORT} || '5432';
  my $dbuser = $ENV{ZING_DBUSER} || 'postgres';
  my $dbpass = $ENV{ZING_DBPASS};
  require DBI; DBI->connect(
    join(';',
      "dbi:Pg:dbname=$dbname",
      $dbhost ? join('=', 'host', $dbhost) : (),
      $dbport ? join('=', 'port', $dbport) : (),
    ),
    $dbuser, $dbpass,
    {
      AutoCommit => 1,
      PrintError => 0,
      RaiseError => 1
    }
  );
}

has meta => (
  is => 'ro',
  isa => 'Str',
  new => 1,
);

fun new_meta($self) {
  require Zing::ID; Zing::ID->new->string
}

has table => (
  is => 'ro',
  isa => 'Str',
  new => 1,
);

fun new_table($self) {
  $ENV{ZING_DBZONE} || 'entities'
}

# BUILDERS

fun new_encoder($self) {
  require Zing::Encoder::Dump; Zing::Encoder::Dump->new;
}

fun BUILD($self) {
  my $client = $self->client;
  my $table = $self->table;
  local $@; eval {
    $client->do(qq{
      create table if not exists "$table" (
        "id" serial primary key,
        "key" varchar not null,
        "value" text not null,
        "index" integer default 0,
        "meta" varchar null
      )
    });
  }
  unless (defined(do{
    local $@;
    local $client->{RaiseError} = 0;
    local $client->{PrintError} = 0;
    eval {
      $client->do(qq{
        select 1 from "$table" where 1 = 1
      })
    }
  }));
  return $self;
}

fun DESTROY($self) {
  $self->client->disconnect;
  return $self;
}

# METHODS

my $retries = 10;

method drop(Str $key) {
  my $table = $self->table;
  my $client = $self->client;
  my $sth = $client->prepare(
    qq{delete from "$table" where "key" = ?}
  );
  $sth->execute($key);
  return $sth->rows > 0 ? 1 : 0;
}

method keys(Str $query) {
  $query =~ s/\*/%/g;
  my $table = $self->table;
  my $client = $self->client;
  my $data = $client->selectall_arrayref(
    qq{select distinct("key") from "$table" where "key" like ?},
    {},
    $query,
  );
  return [map $$_[0], @$data];
}

method lpull(Str $key) {
  my $table = $self->table;
  my $client = $self->client;
  for my $attempt (1..$retries) {
    local $@; eval {
      my $sth = $client->prepare(
        qq{
          update "$table" set "meta" = ? where "id" = (
            select "me"."id" from "$table" "me"
            where "me"."key" = ? and "me"."meta" is null
            order by "me"."index" asc limit 1
          )
        }
      );
      $sth->execute($self->meta, $key);
    };
    if ($@) {
      die $@ if $attempt == $retries;
    }
    else {
      last;
    }
  }
  my $data = $client->selectrow_arrayref(
    qq{
      select "id", "value"
      from "$table" where "meta" = ? and "key" = ? order by "index" asc limit 1
    },
    {},
    $self->meta, $key,
  );
  if ($data) {
    my $sth = $client->prepare(
      qq{delete from "$table" where "id" = ?}
    );
    $sth->execute($data->[0]);
  }
  return $data ? $self->decode($data->[1]) : undef;
}

method lpush(Str $key, HashRef $val) {
  my $table = $self->table;
  my $client = $self->client;
  my $sth = $client->prepare(
    qq{
      insert into "$table" ("key", "value", "index") values (?, ?, (
        select coalesce(min("me"."index"), 0) - 1
        from "$table" "me" where "me"."key" = ?
      ))
    }
  );
  for my $attempt (1..$retries) {
    local $@; eval {
      $sth->execute($key, $self->encode($val), $key);
    };
    if ($@) {
      die $@ if $attempt == $retries;
    }
    else {
      last;
    }
  }
  return $sth->rows;
}

method read(Str $key) {
  my $table = $self->table;
  my $client = $self->client;
  my $data = $client->selectrow_arrayref(
    qq{
      select "value" from "$table"
      where "key" = ? order by "id" desc limit 1
    },
    {},
    $key,
  );
  return $data ? $data->[0] : undef;
}

method recv(Str $key) {
  my $data = $self->read($key);
  return $data ? $self->decode($data) : $data;
}

method rpull(Str $key) {
  my $table = $self->table;
  my $client = $self->client;
  for my $attempt (1..$retries) {
    local $@; eval {
      my $sth = $client->prepare(
        qq{
          update "$table" set "meta" = ? where "id" = (
            select "me"."id" from "$table" "me"
            where "me"."key" = ? and "me"."meta" is null
            order by "me"."index" desc limit 1
          )
        }
      );
      $sth->execute($self->meta, $key);
    };
    if ($@) {
      die $@ if $attempt == $retries;
    }
    else {
      last;
    }
  }
  my $data = $client->selectrow_arrayref(
    qq{
      select "id", "value"
      from "$table" where "meta" = ? and "key" = ? order by "index" desc limit 1
    },
    {},
    $self->meta, $key,
  );
  if ($data) {
    my $sth = $client->prepare(
      qq{delete from "$table" where "id" = ?}
    );
    $sth->execute($data->[0]);
  }
  return $data ? $self->decode($data->[1]) : undef;
}

method rpush(Str $key, HashRef $val) {
  my $table = $self->table;
  my $client = $self->client;
  my $sth = $client->prepare(
    qq{
      insert into "$table" ("key", "value", "index") values (?, ?, (
        select coalesce(max("me"."index"), 0) + 1
        from "$table" "me" where "me"."key" = ?
      ))
    }
  );
  for my $attempt (1..$retries) {
    local $@; eval {
      $sth->execute($key, $self->encode($val), $key);
    };
    if ($@) {
      die $@ if $attempt == $retries;
    }
    else {
      last;
    }
  }
  return $sth->rows;
}

method send(Str $key, HashRef $val) {
  my $set = $self->encode($val);
  $self->write($key, $set);
  return 'OK';
}

method size(Str $key) {
  my $table = $self->table;
  my $client = $self->client;
  my $data = $client->selectrow_arrayref(
    qq{select count("key") from "$table" where "key" = ?},
    {},
    $key,
  );
  return $data->[0];
}

method slot(Str $key, Int $pos) {
  my $table = $self->table;
  my $client = $self->client;
  my $data = $client->selectrow_arrayref(
    qq{
      select "value" from "$table"
      where "key" = ? order by "index" asc offset ? limit 1
    },
    {},
    $key, $pos
  );
  return $data ? $self->decode($data->[0]) : undef;
}

method test(Str $key) {
  my $table = $self->table;
  my $client = $self->client;
  my $data = $client->selectrow_arrayref(
    qq{select count("id") from "$table" where "key" = ?},
    {},
    $key,
  );
  return $data->[0] ? 1 : 0;
}

method write(Str $key, Str $data) {
  my $table = $self->table;
  my $client = $self->client;
  $client->prepare(
    qq{delete from "$table" where "key" = ?}
  )->execute($key);
  $client->prepare(
    qq{insert into "$table" ("key", "value") values (?, ?)}
  )->execute($key, $data);
  return $self;
}

1;
